feat(processing): Extract event timestamp from minidump

flub opened the pull request:
If a minidump event is received, this extracts the timestamp from the
minidump and sets it as the event's timestamp.

Commit:
Extract event timestamp from minidump

If a minidump event is received, this extracts the timestamp from the
minidump and sets it as the event's timestamp.

Swatinem commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+    fn write_minidump_timestamp(event: &mut Event, minidump_item: &envelope::Item) {
+        use std::time::SystemTime;
+
+        assert_eq!(
```
do we have such asserts already? This is clearly a programmer error, so using panic in this case seems okay, or maybe use the typestate pattern to make this impossible to call?

untitaker replied:
our existing pattern is to use dbg_assert, we very unlikely want to panic in prod

jan-auer replied:
Definitely do not assert and rather return an error (or swallow it silently). Given how the other functions here are coded (e.g. `parse_msgpack_breadcrumbs`), I would just early return.

Swatinem commented on relay-server/src/actors/events.rs:
```diff
@@ -769,12 +804,14 @@ impl EventProcessor {
             state.metrics.bytes_ingested_event_minidump = Annotated::new(item.len() as u64);
-            self.write_native_placeholder(&mut state.event, true);
+            self.write_native_placeholder(&mut event, true);
+            Self::write_minidump_timestamp(&mut event, item);
```
maybe just make it a free function then? Its a bit unconventional to have static functions that are not constructors.
Maybe instead of taking a `&mut Event`, rename the function to `try_extract_timestamp` or similar and return an `Option<DateTime<Utc>>`.

jan-auer replied:
Since all functions are on `EventProcessor` for the time, I'd suggest to just give it an unused `&self`. In one of the upcoming changes we will likely have to access the `config`, so it's totally fine to have it on there. I think taking `&mut Event` is fine in this case.

untitaker commented on relay-server/src/actors/events.rs:
```diff
@@ -50,6 +51,17 @@ use {
+/// Counter for minidump parsing.
+///
+/// Emits metrics under the "minidump.parse" name.
+struct MinidumpParseCounter;
```
Would move this into metrics module to the other impls for consistency

untitaker replied:
Also if you move it there the docs should auto-update: https://getsentry.github.io/relay/configuration/metrics/#eventqueue_sizepct

jan-auer replied:
What Markus said + We probably don't need this metric at all, there is no precedent for logging metrics on individual attachments at the moment. I would prefer if we remove this metric, and handle this consistently with other special attachments.

untitaker commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+        let minidump = match minidump::Minidump::read(cursor) {
+            Ok(minidump) => minidump,
+            Err(err) => {
+                log::error!("Failed to parse minidump: {:?}", err);
```
iirc this will create a sentry event to our internal projects, do we want that?

jan-auer replied:
Agreed with this. Please only log this on `debug!` like in the other cases and remove the metric.

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+            }
+        };
+        let epoch_offset = Duration::from_secs(minidump.header.time_date_stamp.into());
+        let when: DateTime<Utc> = DateTime::from(SystemTime::UNIX_EPOCH + epoch_offset);
```
You can simplify this using [`TimeZone::timestamp`](https://docs.rs/chrono/0.4.13/chrono/offset/trait.TimeZone.html#method.timestamp):

```suggestion
        let timestamp = Utc::timestamp(minidump.header.time_date_stamp.into(), 0);
```

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+            minidump_item.content_type(),
+            Some(&envelope::ContentType::Minidump)
+        );
+        let cursor = std::io::Cursor::new(minidump_item.payload());
```
Please import `Cursor`.

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+    /// Extracts the timestamp from the minidump and uses it as the event timestamp.
+    #[cfg(feature = "processing")]
+    fn write_minidump_timestamp(event: &mut Event, minidump_item: &envelope::Item) {
+        use std::time::SystemTime;
```
Please move the import to the top of the file. However, I think you can remove it based on a comment below.

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
 
+    /// Extracts the timestamp from the minidump and uses it as the event timestamp.
+    #[cfg(feature = "processing")]
+    fn write_minidump_timestamp(event: &mut Event, minidump_item: &envelope::Item) {
```
Please import `Item`.

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -754,6 +764,31 @@ impl EventProcessor {
+        let epoch_offset = Duration::from_secs(minidump.header.time_date_stamp.into());
+        let when: DateTime<Utc> = DateTime::from(SystemTime::UNIX_EPOCH + epoch_offset);
+        let event_timestamp = event.timestamp.value_mut();
+        *event_timestamp = Some(when);
```
```suggestion
        event.timestamp.set_value(Some(when));
```

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -769,12 +804,14 @@ impl EventProcessor {
         let apple_crash_report_attachment = envelope
             .get_item_by(|item| item.attachment_type() == Some(AttachmentType::AppleCrashReport));
 
+        let mut event = state.event.get_or_insert_with(Event::default);
```
Move this into the `if let Some()` because we must not create the event payload if there is no attachment.

jan-auer commented on relay-server/Cargo.toml:
```diff
@@ -59,6 +59,7 @@ symbolic = { version = "7.4.0", optional = true, default-features=false, feature
 tokio-timer = "0.2.13"
 url = { version = "2.1.1", features = ["serde"] }
 uuid = { version = "0.8.1", features = ["v5"] }
+minidump = "0.1.0"
```
_nit_: Sort dependencies alphabetically.

jan-auer reviewed (changes requested):
Thanks!

Commit:
Make minidump crate a conditional dependency

Commit:
Do not yet log metrics, avoid logging errors

We want to do the counters separately in a followup.

Logging errors shows up as errors in sentry, so avoid doing this.

Commit:
Some minor cleanup from feedback

Commit:
Fix testing code

This is why testing code should have tests.  Still not adding them
because I like to hurt myself some more first in the misguided believe
this will save me time.

Commit:
Use git dependency on minidump crate

There have been some changes and the functionality is much better.
This allows us to avoid the Cursor wrapping which is slow in debug
builds.

jan-auer committed:
test: Extract timestamp from minidump

jan-auer committed:
Merge branch 'master' into feat/minidump-timestamp

* master:
  doc: Clarify that managed mode is not GA
  release: 20.7.2
  ref(server): Increase maximum attachment size to 100MiB (#671)
  ref(config): Clear up confusion around mebibytes (#670)
  meta: Changelog
  feat(server): Report metrics for upstream connections (#669)
  test: Fix flakey outcomes test (#668)
  ref(transactions): Normalize transaction spans into an array (#667)
  release: 20.7.1

jan-auer committed:
meta: Changelog

jan-auer committed:
build: Pin minidump, fix warning

jan-auer reviewed (approved):
Thanks a bunch!

cc @untitaker: this allows us to remove [`validate_and_set_timestamp`](https://github.com/getsentry/sentry/blob/42c170c4a61352db969b8ef811b0802016a076e7/src/sentry/event_manager.py#L101) in Sentry.
